Scale/units conversion

ericpre opened the pull request:
This PR supersedes #1106.
- add the `convert_to_units` method to `DataAxis` (can be inplace or not)
- add the `convert_units` method to `AxisManager`
- support slicing with units
- by default, the units are converted automatically when loading
- add `scale_as_quantity` and `offset_as_quantity` properties.

Examples:
```python
>>> import hyperspy.api as hs
>>> axis = hs.hyperspy.axes.DataAxis(size=10, scale=0.1, offset=10, units='mm')
>>> print("scale: {}, offset: {}, units: {}".format(axis.scale, axis.offset, axis.units))
>>> axis.convert_to_units('um')
>>> print("scale: {}, offset: {}, units: {}".format(axis.scale, axis.offset, axis.units))
scale: 0.1, offset: 10.0, units: mm
scale: 100.0, offset: 10000.0, units: µm
>>> axis.convert_to_units('cm', inplace=False)
(100.00000000000001, 10000.0, 'um')
```

```python
>>> import numpy as np
>>> s = hs.signals.Signal1D(np.random.random((50, 50, 2048)))
>>> am = s.axes_manager
>>> for axis in am.navigation_axes:
>>>     axis.units = 'm'
>>>     axis.scale = 3e-9
>>> for axis in am.signal_axes:
>>>     axis.units = 'eV'
>>>     axis.scale = 1.e-3
>>> am2 = am.deepcopy()
>>> am
<Axes manager, axes: (50, 50|2048)>
            Name |   size |  index |  offset |   scale |  units 
================ | ====== | ====== | ======= | ======= | ====== 
     <undefined> |     50 |      0 |       0 |   3e-09 |      m 
     <undefined> |     50 |      0 |       0 |   3e-09 |      m 
---------------- | ------ | ------ | ------- | ------- | ------ 
     <undefined> |   2048 |        |       0 |  0.0015 |     eV 

>>> am.convert_units()
>>> am
<Axes manager, axes: (50, 50|2048)>
            Name |   size |  index |  offset |   scale |  units 
================ | ====== | ====== | ======= | ======= | ====== 
     <undefined> |     50 |      0 |       0 |       3 |     nm 
     <undefined> |     50 |      0 |       0 |       3 |     nm 
---------------- | ------ | ------ | ------- | ------- | ------ 
     <undefined> |   2048 |        |       0 |     1.5 |    meV 

>>> am2.convert_units('signal')
<Axes manager, axes: (50, 50|2048)>
            Name |   size |  index |  offset |   scale |  units 
================ | ====== | ====== | ======= | ======= | ====== 
     <undefined> |     50 |      0 |       0 |   3e-09 |      m 
     <undefined> |     50 |      0 |       0 |   3e-09 |      m 
---------------- | ------ | ------ | ------- | ------- | ------ 
     <undefined> |   2048 |        |       0 |     1.5 |    meV 

>>> am.convert_units('navigation', 'um')
>>> am
<Axes manager, axes: (50, 50|2048)>
            Name |   size |  index |  offset |   scale |  units 
================ | ====== | ====== | ======= | ======= | ====== 
     <undefined> |     50 |      0 |       0 |   0.003 |     µm 
     <undefined> |     50 |      0 |       0 |   0.003 |     µm 
---------------- | ------ | ------ | ------- | ------- | ------ 
     <undefined> |   2048 |        |       0 |  0.0015 |     eV 
```

``` python
>>> s = hs.signals.Signal1D(np.arange(10))
>>> s.axes_manager[0].units = 'µm'
>>> s.isig[:'2000 nm'].data
array([0, 1, 2, 3])
```

``` python
>>> s = hs.signals.Signal1D(np.arange(10))
>>> s.axes_manager[0].scale_as_quantity = '2.5 µm'
>>> s.axes_manager
<Axes manager, axes: (|10)>
            Name |   size |  index |  offset |   scale |  units 
================ | ====== | ====== | ======= | ======= | ====== 
---------------- | ------ | ------ | ------- | ------- | ------ 
     <undefined> |     10 |        |       0 |     2.5 |     µm 
>>> s.axes_manager[0].offset_as_quantity = '2.5 nm'
<Axes manager, axes: (|10)>
            Name |   size |  index |  offset |   scale |  units 
================ | ====== | ====== | ======= | ======= | ====== 
---------------- | ------ | ------ | ------- | ------- | ------ 
     <undefined> |     10 |        |     2.5 | 2.5e+03 |     nm 
```

ericpre committed:
add pint>=0.7 dependency

ericpre committed:
add convert_to_units method to axes_manager using pint to_compact  method and corresponding tests.

ericpre committed:
move unit conversion functions to UnitConversion class

ericpre committed:
Add auto_convert_units argument to load and crop. Default set to do nothing.

ericpre committed:
add auto_convert_units to load

ericpre committed:
convert units during plotting (work in progress)

ericpre committed:
Merge remote-tracking branch 'upstream/RELEASE_next_minor' into scale_unit_conversion

# Conflicts:
#	hyperspy/tests/io/test_tiff.py

ericpre committed:
fix plot auto_convert_units argument

ericpre committed:
convert also offset and rename tests file

ericpre committed:
convert units navigators

ericpre committed:
add axes_off (default=false) argument for plotting images to remove axes label

ericpre committed:
Merge remote-tracking branch 'upstream/RELEASE_next_minor' into scale_unit_conversion

# Conflicts:
#	.travis.yml
#	anaconda_hyperspy_environment.yml
#	appveyor.yml
#	hyperspy/axes.py
#	hyperspy/drawing/image.py
#	hyperspy/drawing/mpl_hse.py
#	hyperspy/io.py
#	hyperspy/signal.py
#	setup.py

ericpre committed:
Fix tests after merge.

ericpre committed:
Set convert_units to True, when loading file and fix corresponding tests

ericpre committed:
Update plot testing baseline images.

ericpre committed:
Simplify code and fix test

ericpre committed:
Remove warning and use the first axis units when several axis have different converted units.

ericpre commented:
I initially tried to convert the units only when plotting (without changing the `axes_manager` of the signal) and I was doing it by copying the `axes_manager` at the creation of the plot. For obvious reason it was not possible anymore to change the position of the pointer from the signal `axes_manager`, since both axes_manager were different. It could be possible to find workaround this, but I think that this more simple the way how it is now: converting automatically at loading or cropping.

ericpre committed:
update baseline images for plot testing.

ericpre committed:
travis ci: use natsort from anaconda channel rather than conda-forge to have a more recent version and avoid warning interference.

francisco-dlp commented:
Very cool! Could you add an example showing how these new features work in practice to the original comment? I think that this is good practice as it makes it easier to understand what the PR does for users and reviewers (we should add it to the Dev Guide).

I think that automatically changing the units is dangerous. For example, to perform math operation on the data the units have to be consistent and it's going to be difficult to track when/where the units have changed if hspy does it automatically. We could check the units when performing math operations on signals but it is also common to perform math operations directly using the numpy array, so even that wouldn't fix the issue.

ericpre commented on hyperspy/signal.py:
```diff
@@ -1928,7 +1928,7 @@ def __call__(self, axes_manager=None):
 
     def plot(self, navigator="auto", axes_manager=None,
-             plot_markers=True, **kwargs):
+             convert_units=True, plot_markers=True, **kwargs):
```
Actually, if the units are converted during loading, there is no point to set `convert_units=True` as default.

ericpre commented:
> Could you add an example showing how these new features work in practice to the original comment? I think that this is good practice as it makes it easier to understand what the PR does for users and reviewers (we should add it to the Dev Guide).

Maybe, we should set a template for issue and pull request? See [https://github.com/blog/2111-issue-and-pull-request-templates](https://github.com/blog/2111-issue-and-pull-request-templates).

francisco-dlp commented:
Sounds good for PRs. For issues it's a bit harder because people use issues to report bugs, ask for help with a certain problem, ask for features and having to fill a general template that doesn't match with the user topic might work as an (unwanted) deterrent.

ericpre committed:
Fix bug, improve docstring and add a few more tests.

ericpre committed:
Add slicing with units with tests and docs.

ericpre committed:
Update doc.

ericpre committed:
Add `scale_as_quantity` and `offset_as_quantity` properties to `DataAxis`.

ericpre committed:
Fix bad design with `scale_as_quantity` property.

ericpre committed:
Fix bug with `scale_as_quantity` property.

to266 commented on doc/user_guide/tools.rst:
```diff
@@ -829,7 +832,7 @@ using an external function can be more easily accomplished using the
   Rotation of images by the same amount using :py:meth:`~.signal.BaseSignal.map`.
 
-The :py:meth:`~.signal.BaseSignal.map` method can also take variable
+The :py:meth:`~.signal.Bfancy inaseSignal.map` method can also take variable
```
Not required

to266 commented on hyperspy/axes.py:
```diff
@@ -825,6 +990,98 @@ def _on_scale_changed(self):
+        elif axes == 'signal':
+            axes = self.signal_axes
+            convert_navigation = False
+        elif isinstance(axes, DataAxis) or type(axes) is int or type(axes) is str:
```
`isinstance(axes, (DataAxis, int, str))` should do the same

to266 commented on hyperspy/axes.py:
```diff
@@ -825,6 +990,98 @@ def _on_scale_changed(self):
+                axes = self[axes]
+            axes = (axes, )
+            convert_navigation = axes[0].navigate
+            convert_signal = not axes[0].navigate
```
I would prefer
`conver_signal = not convert_navigation`
it's both faster (no `axes[0]` lookup) and clearer that the two are opposite

to266 commented on hyperspy/axes.py:
```diff
@@ -825,6 +990,98 @@ def _on_scale_changed(self):
+            raise TypeError(
+                'Axes type `{}` is not correct.'.format(type(axes)))
+
+        if type(units) is str or units is None:
```
`isinstance(units, str)`

to266 commented on hyperspy/axes.py:
```diff
@@ -825,6 +990,98 @@ def _on_scale_changed(self):
+
+        if type(units) is str or units is None:
+            units = [units] * len(axes)
+        elif type(units) is list:
```
`if isinstance(units, list) and len(units) != len(axes):`

i.e. combine the two conditionals

to266 replied:
Alternatively you could just loop the last one if not enough units are given:
```python
units = ['um', 'nm']
axes = [1,2,3]
units = units + units[-1:]*(len(axes) - len(units))
```

ericpre replied:
Combining both conditionals is not working here, because it will skip the case when units is a list with the correct length and it will end up raising an error in the `else` block.

I am a bit sceptical about extrapolating the list of units, for the case when navigation and signals axis are passed. I would prefer to force the user to specify the correct units.

to266 commented on hyperspy/axes.py:
```diff
@@ -825,6 +990,98 @@ def _on_scale_changed(self):
+                        axis.convert_to_units(axes[0].units, factor)
+                    return
+            else:
+                current_units = axis.units  # initialisation first iteration
```
If you only intend to keep the first `is not None and len(unit)` unit, why not find it first from the list, and then loop over the axes just once? 
If this seems acceptable, how about this:
```python
first_good_unit = next(filter(lambda x: x is not None and x, units))
# Change `micron` to `um` if you want here as well
for axis in filter(lambda ax: ax.unit != first_good_unit, axes):
    axis.convert_to_units(first_good_unit, factor)
```

ericpre replied:
I simplified here, actually it was very simple and really unnecessary complicated!!

to266 reviewed (changes requested):
A quick look through the code. Once corrected, will need another check

ericpre committed:
Address PR review comments.

ericpre committed:
Simplify `_convert_axes_to_same_units` for different axes.

ericpre committed:
Fix a few issues after simplification code.

ericpre committed:
Add the option to convert the axis units "inplace" or not.

ericpre committed:
Merge remote-tracking branch 'upstream/RELEASE_next_minor' into scale_unit_conversion_take2

# Conflicts:
#	.travis.yml
#	hyperspy/drawing/image.py
#	hyperspy/io.py
#	hyperspy/tests/drawing/plot_signal1d/test_plot_two_cursors_1-sig.png
#	hyperspy/tests/drawing/plot_signal1d/test_plot_two_cursors_2-sig.png
#	hyperspy/tests/io/test_edax.py
#	hyperspy/tests/io/test_emd.py
#	hyperspy/tests/io/test_tiff.py

ericpre committed:
Fix tests, one bug and add more tests.

ericpre committed:
Merge remote-tracking branch 'upstream/RELEASE_next_minor' into HEAD

# Conflicts:
#	hyperspy/tests/drawing/plot_markers/test_plot_line_markers.png
#	hyperspy/tests/drawing/plot_markers/test_plot_point_markers.png
#	hyperspy/tests/drawing/plot_markers/test_plot_rectange_markers.png
#	hyperspy/tests/drawing/plot_markers/test_plot_text_markers_sig.png
#	hyperspy/tests/drawing/plot_signal/test_plot_sig_nav_0-1-sig-complex_imag.png
#	hyperspy/tests/drawing/plot_signal/test_plot_sig_nav_0-1-sig-complex_real.png
#	hyperspy/tests/drawing/plot_signal/test_plot_sig_nav_0-1-sig-real.png
#	hyperspy/tests/drawing/plot_signal/test_plot_sig_nav_0-2-sig-complex_imag.png
#	hyperspy/tests/drawing/plot_signal/test_plot_sig_nav_0-2-sig-complex_real.png
#	hyperspy/tests/drawing/plot_signal/test_plot_sig_nav_0-2-sig-real.png
#	hyperspy/tests/drawing/plot_signal/test_plot_sig_nav_1-1-nav-complex_imag.png
#	hyperspy/tests/drawing/plot_signal/test_plot_sig_nav_1-1-nav-complex_real.png
#	hyperspy/tests/drawing/plot_signal/test_plot_sig_nav_1-1-nav-real.png
#	hyperspy/tests/drawing/plot_signal/test_plot_sig_nav_1-1-sig-complex_imag.png
#	hyperspy/tests/drawing/plot_signal/test_plot_sig_nav_1-1-sig-complex_real.png
#	hyperspy/tests/drawing/plot_signal/test_plot_sig_nav_1-1-sig-real.png
#	hyperspy/tests/drawing/plot_signal/test_plot_sig_nav_1-2-nav-complex_imag.png
#	hyperspy/tests/drawing/plot_signal/test_plot_sig_nav_1-2-nav-complex_real.png
#	hyperspy/tests/drawing/plot_signal/test_plot_sig_nav_1-2-nav-real.png
#	hyperspy/tests/drawing/plot_signal/test_plot_sig_nav_1-2-sig-complex_imag.png
#	hyperspy/tests/drawing/plot_signal/test_plot_sig_nav_1-2-sig-complex_real.png
#	hyperspy/tests/drawing/plot_signal/test_plot_sig_nav_1-2-sig-real.png
#	hyperspy/tests/drawing/plot_signal/test_plot_sig_nav_2-1-nav-complex_imag.png
#	hyperspy/tests/drawing/plot_signal/test_plot_sig_nav_2-1-nav-complex_real.png
#	hyperspy/tests/drawing/plot_signal/test_plot_sig_nav_2-1-nav-real.png
#	hyperspy/tests/drawing/plot_signal/test_plot_sig_nav_2-1-sig-complex_imag.png
#	hyperspy/tests/drawing/plot_signal/test_plot_sig_nav_2-1-sig-complex_real.png
#	hyperspy/tests/drawing/plot_signal/test_plot_sig_nav_2-1-sig-real.png
#	hyperspy/tests/drawing/plot_signal/test_plot_sig_nav_2-2-nav-complex_imag.png
#	hyperspy/tests/drawing/plot_signal/test_plot_sig_nav_2-2-nav-complex_real.png
#	hyperspy/tests/drawing/plot_signal/test_plot_sig_nav_2-2-nav-real.png
#	hyperspy/tests/drawing/plot_signal/test_plot_sig_nav_2-2-sig-complex_imag.png
#	hyperspy/tests/drawing/plot_signal/test_plot_sig_nav_2-2-sig-complex_real.png
#	hyperspy/tests/drawing/plot_signal/test_plot_sig_nav_2-2-sig-real.png
#	hyperspy/tests/drawing/plot_signal1d/test_plot_spectra_default.png
#	hyperspy/tests/drawing/plot_signal1d/test_plot_spectra_heatmap.png
#	hyperspy/tests/drawing/plot_signal1d/test_plot_spectra_mosaic.png
#	hyperspy/tests/drawing/plot_signal1d/test_plot_spectra_overlap.png
#	hyperspy/tests/drawing/plot_signal1d/test_plot_spectra_sync_1nav.png
#	hyperspy/tests/drawing/plot_signal1d/test_plot_two_cursors_1-nav.png
#	hyperspy/tests/drawing/plot_signal1d/test_plot_two_cursors_2-nav.png
#	hyperspy/tests/drawing/plot_signal2d/test_plot_False-False-False-False.png
#	hyperspy/tests/drawing/plot_signal2d/test_plot_False-False-False-True.png
#	hyperspy/tests/drawing/plot_signal2d/test_plot_False-False-True-False.png
#	hyperspy/tests/drawing/plot_signal2d/test_plot_False-False-True-True.png
#	hyperspy/tests/drawing/plot_signal2d/test_plot_False-True-False-False.png
#	hyperspy/tests/drawing/plot_signal2d/test_plot_False-True-False-True.png
#	hyperspy/tests/drawing/plot_signal2d/test_plot_False-True-True-False.png
#	hyperspy/tests/drawing/plot_signal2d/test_plot_False-True-True-True.png
#	hyperspy/tests/drawing/plot_signal2d/test_plot_True-False-False-False.png
#	hyperspy/tests/drawing/plot_signal2d/test_plot_True-False-False-True.png
#	hyperspy/tests/drawing/plot_signal2d/test_plot_True-False-True-False.png
#	hyperspy/tests/drawing/plot_signal2d/test_plot_True-False-True-True.png
#	hyperspy/tests/drawing/plot_signal2d/test_plot_True-True-False-False.png
#	hyperspy/tests/drawing/plot_signal2d/test_plot_True-True-False-True.png
#	hyperspy/tests/drawing/plot_signal2d/test_plot_True-True-True-False.png
#	hyperspy/tests/drawing/plot_signal2d/test_plot_True-True-True-True.png
#	hyperspy/tests/drawing/plot_signal2d/test_plot_multiple_images_list_None-None.png
#	hyperspy/tests/drawing/plot_signal2d/test_plot_multiple_images_list_vmin0-vmax0.png

ericpre committed:
Update baseline images for plotting tests.

ericpre committed:
Add io plugin characteristics to set the automatic conversion of units. Improve load docstring and documentation.

ericpre commented:
@to266, would be possible for you to review this again?

k8macarthur commented on hyperspy/axes.py:
```diff
@@ -56,13 +61,142 @@ def generate_axis(offset, scale, size, offset_index=0):
+    def _convert_compact_units(self, factor=0.25, inplace=True):
+        """ Convert units to "human-readable" units, which means with a 
+            convenient prefix. `factor` is a adjustage factor used to 
+            determine the prefix of the units (in combinaison with the size of 
```
*combination

k8macarthur commented on hyperspy/axes.py:
```diff
@@ -56,13 +61,142 @@ def generate_axis(offset, scale, size, offset_index=0):
+        """ Convert the scale and the units of the current axis. If the units
+        is not supported by the pint library, the scale and units are not
+        changed.
+
```
If the unit of measure is not supported....

k8macarthur commented on hyperspy/axes.py:
```diff
@@ -56,13 +61,142 @@ def generate_axis(offset, scale, size, offset_index=0):
+        units : {str | None}
+            Default = None
+            If str, the axis will converted to the provided units.
+            If `None`, the scale and the units are converted to the appropriate 
```
If str, the axis will be converted to the provided units.

k8macarthur commented on hyperspy/axes.py:
```diff
@@ -56,13 +61,142 @@ def generate_axis(offset, scale, size, offset_index=0):
+            If `True`, convert the axis in place. if `False` return the  
+            `scale`, `offset` and `units`.
+        factor : float
+            `factor` is a adjustage factor used to determine the prefix of the 
```
'factor' an adjustable value used to determine the prefix of the units.

k8macarthur commented on hyperspy/axes.py:
```diff
@@ -56,13 +61,142 @@ def generate_axis(offset, scale, size, offset_index=0):
+            `scale`, `offset` and `units`.
+        factor : float
+            `factor` is a adjustage factor used to determine the prefix of the 
+            units (in combinaison with the size of the axis).
```
*combination

k8macarthur commented on hyperspy/axes.py:
```diff
@@ -825,6 +997,103 @@ def _on_scale_changed(self):
+    def convert_units(self, axes=None, units=None, same_units=True,
+                      factor=0.25):
+        """ Convert the scale and the units of the selected axes. If the units
+        is not supported by the pint library, the scale and units are not
```
If the unit of measure is not supported....

k8macarthur commented on hyperspy/axes.py:
```diff
@@ -825,6 +997,103 @@ def _on_scale_changed(self):
+        ----------
+        axes : {int | string | iterable of `DataAxis` | None}
+            Default = None
+            Convert to the convenient scale and units on the specified axis.
```
Convert to a convenient scale

k8macarthur commented on hyperspy/axes.py:
```diff
@@ -825,6 +997,103 @@ def _on_scale_changed(self):
+        units : {list of string of the same length than axes | str | None}
+            Default = None
+            If list, the selected axes will be converted to the provided units.
+            If str, the navigation or signal axes will converted to the 
```
will be converted

k8macarthur commented on hyperspy/axes.py:
```diff
@@ -825,6 +997,103 @@ def _on_scale_changed(self):
+            first axis is used for all axes. If `False`, convert all axes 
+            individually.
+        factor : float
+            `factor` is a adjustage factor used to determine the prefix of the 
```
'factor' an adjustable value used to determine the prefix of the units.

k8macarthur commented on hyperspy/io_plugins/blockfile.py:
```diff
@@ -107,23 +107,29 @@ def get_header_from_signal(signal, endianess='<'):
-    elif signal.axes_manager.navigation_dimension == 1:
-        NX = signal.axes_manager.navigation_shape[0]
+    # The navigation and signal units are 'nm' and 'cm', respectively, so we
+    # convert the units according before saving the signal
```
accordingly

k8macarthur commented:
Found a few typos in the docstrings.

ericpre committed:
Merge remote-tracking branch 'upstream/RELEASE_next_minor' into scale_unit_conversion_take2

# Conflicts:
#	hyperspy/io.py
#	hyperspy/io_plugins/bruker.py
#	hyperspy/tests/io/test_emd.py
#	hyperspy/tests/signal/test_tools.py

ericpre committed:
Fix typos, thanks @k8macarthur.

francisco-dlp commented:
@ericpre, could you merge Rnm into this?

ericpre committed:
Merge remote-tracking branch 'upstream/RELEASE_next_minor' into scale_unit_conversion_take2

# Conflicts:
#	hyperspy/axes.py
#	hyperspy/tests/drawing/plot_signal1d/test_plot_spectra_mosaic.png
#	hyperspy/tests/signal/test_tools.py

francisco-dlp commented on hyperspy/axes.py:
```diff
@@ -57,13 +63,142 @@ def generate_axis(offset, scale, size, offset_index=0):
+    def convert_to_units(self, units=None, inplace=True, factor=0.25):
+        """ Convert the scale and the units of the current axis. If the unit 
+        of measure is not supported by the pint library, the scale and units 
+        are not changed.
```
changed -> modified?

francisco-dlp commented on hyperspy/axes.py:
```diff
@@ -57,13 +63,142 @@ def generate_axis(offset, scale, size, offset_index=0):
+            Default = None
+            If str, the axis will be converted to the provided units.
+            If `None`, the scale and the units are converted to the appropriate 
+            scale and units to avoid displaying scalebar with >3 digits or too 
```
What about changing ``None`` to ``"auto"``?

Simpler docstring: If "auto", automatically determine the optimal units too big or too small numbers.

francisco-dlp commented on hyperspy/axes.py:
```diff
@@ -57,13 +63,142 @@ def generate_axis(offset, scale, size, offset_index=0):
+        inplace : bool
+            If `True`, convert the axis in place. if `False` return the  
+            `scale`, `offset` and `units`.
+        factor : float
```
It would be good to give more details on how ``factor`` affects the automatic units conversion.

francisco-dlp commented on hyperspy/axes.py:
```diff
@@ -57,13 +63,142 @@ def generate_axis(offset, scale, size, offset_index=0):
+        elif s == '':
+            self._units = t.Undefined
+        else:
+            self._units = s.replace('um', 'µm').replace(' ', '')
```
I think that it'll be better to avoid preprocessing the string—if that was a good idea pint would do it.

ericpre replied:
I have opened an issue in the [pint repo](https://github.com/hgrecco/pint/issues/666) to ask. Pint seems to have a fairly good unicode support, so I don't see, why it would not work. Maybe, to avoid issue with python < 2.6?

francisco-dlp replied:
Personally I would prefer not to perform this translation in HyperSpy as it can lead to unexpected issues. For example, somebody may define an unit that contains the "um" chain. Also, it is not so hard to write μ even I just did it!

ericpre replied:
Yes, indeed, this is not right, I revert it.

francisco-dlp commented on hyperspy/axes.py:
```diff
@@ -57,13 +63,142 @@ def generate_axis(offset, scale, size, offset_index=0):
                        size)
 
 
+class UnitConversion(object):
```
``object`` not needed in Python3

francisco-dlp commented on hyperspy/axes.py:
```diff
@@ -462,6 +606,35 @@ def update_from(self, axis, attributes=["scale", "offset", "units"]):
         return any_changes
 
+    @property
+    def scale_as_quantity(self):
```
These new properties are not documented in the User Guide.

francisco-dlp commented on hyperspy/axes.py:
```diff
@@ -462,6 +606,35 @@ def update_from(self, axis, attributes=["scale", "offset", "units"]):
+        elif s == '':
+            self._units = t.Undefined
+        else:
+            self._units = s.replace('um', 'µm').replace(' ', '')
```
Same as above.

francisco-dlp commented on hyperspy/axes.py:
```diff
@@ -826,6 +999,103 @@ def _on_scale_changed(self):
+            `navigation` or `signal`. By default the converted units of the 
+            first axis is used for all axes. If `False`, convert all axes 
+            individually.
+        factor : float
```
As ``factor`` appears in many docstrings, it could be extracted into a single string and used where needed.

francisco-dlp commented on hyperspy/io_plugins/blockfile.py:
```diff
@@ -38,7 +38,7 @@
 file_extensions = ['blo', 'BLO']
 default_extension = 0
-
+auto_convert_units = True
```
This does not appear to be used anywhere

francisco-dlp commented on hyperspy/io.py:
```diff
@@ -105,13 +106,16 @@ def load(filenames=None,
-        disk until required. Allows opening arbitrary-sized datasets. default
+        disk until required. Allows opening arbitrary-sized datasets. The default
         is `False`.
+    convert_units : {None, bool}
```
I am not sure if it is a good idea to let the reader decide if to convert units or not. I would rather default to ``False`` and, if the user finds that the units are not convenient, he can always call the ``AxesManager.convert_units`` manually.

ericpre replied:
This is actually one of the motivation of this PR: for example, the fei reader convert scale and offset systematically to "nm" for convenience, because the file is saved with the units in "m".  Same for the bruker reader, the units are saved in "um" in the file, which is not great for nanometer scale EDS mapping. 
Allowing to have a reader parameter to determine if the units needs to be converted or not is an improvement to the current situation. Only the reader, which present this issue have `auto_convert_units=True`.

francisco-dlp replied:
I still think that it is better to default to "don't autoconvert units" and not to hardwire any different behaviour as this is not transparent for the user. Units auto conversion is convenient, but it is somehow unpredictable: it can change with different hyperspy and pint versions and and files stored in the same format could return different units depending on the data stored—too much unpredictability IMO.

Personally I think that the units in the file must be respected i.e. if FEI writes everything in S.I. we should read the data in S.I. by default. I know that this is not the current situations, and for me this fixes it as long as it respects the file units by default. The reason why some file readers automatically convert units is that there was no easy way of converting units before you implemented this feature.

ericpre replied:
Since it is only the FEI reader and some tiff files which can generate inappropriate units, I removed the auto_convert_units reader parameter and I reverted to the previous conversion in the FEI reader, since I would be annoyed to have to write `convert_units=True` each time I want to load an emi/ser file.

francisco-dlp commented on hyperspy/io_plugins/README:
```diff
@@ -13,6 +13,7 @@ All the read/write plugins must provide a python file containing:
         # Recognised file extension
         file_extensions = <Tuple of string>
         default_extension = <Int>	# Index of the extension that will be used by default
+        auto_convert_units = <Bool> # Convert automatically the units
```
As above, I think that it'll be better not to hardwire this option in the reader.

francisco-dlp commented on hyperspy/signal.py:
```diff
@@ -1971,6 +1973,7 @@ def plot(self, navigator="auto", axes_manager=None,
                 "'s.transpose(signal_axes=(1,2)).plot()' "
                 "for plotting as a 2D signal.")
 
+        self._plot.auto_convert_units = convert_units
```
What's the use of this?

francisco-dlp commented on hyperspy/signal.py:
```diff
@@ -1951,6 +1951,8 @@ def plot(self, navigator="auto", axes_manager=None,
 
         if axes_manager is None:
             axes_manager = self.axes_manager
+        if convert_units:
```
I don't think that it is a good idea to add this option to ``plot``: the user may not be aware that this is actually changing the AxesManager units in-place. If that's what the user wants to do, he could call ``AxesManager.concert_units`` manually before calling plot.

It would be possible to change the units only for the plot axes_manager, but that would lead to issues with the ROIs.

ericpre replied:
Yes, it is removed now.

francisco-dlp commented on doc/user_guide/tools.rst:
```diff
@@ -453,6 +453,7 @@ features differ from numpy):
 
```
I suggest to add a section to this chapter of the User Guide explaining how AxesManager manages the units and the new ``convert_units`` ``AxesManager`` and ``DataAxis`` methods.

francisco-dlp reviewed (changes requested):
This is a very welcome new feature. I've left some few comments that need to be addressed before merging.

francisco-dlp commented:
On a side note: would it be possible (in future pull requests) to:

* Force some objects (e.g. ROIs) to always operate on SI units regardless of the AxesManager units to make it possible to operate with the same ROI on axes with different units
* Add units in this way to all metadata numbers and method/function parameters and convert to the right units internally before performing the operations?

ericpre committed:
@francisco-dlp review:
- remove convert_units from the plot method.
- improve docstring.
- add section in the user guide.

ericpre commented:
I have addressed all comments.

> On a side note: would it be possible (in future pull requests) to:
> 
> - Force some objects (e.g. ROIs) to always operate on SI units regardless of the AxesManager units to make it possible to operate with the same ROI on axes with different units
> - Add units in this way to all metadata numbers and method/function parameters and convert to the right units internally before performing the operations?

Yes, it is sensible and since this PR already changes many file, better to do in another one, as you suggested!

ericpre committed:
revert formatting um into µm.

ericpre committed:
Remove auto_convert_units reader parameter

ericpre committed:
Update baseline images for plot testing.

ericpre committed:
Merge remote-tracking branch 'upstream/RELEASE_next_minor' into scale_unit_conversion_take2

francisco-dlp commented on doc/user_guide/io.rst:
```diff
@@ -75,6 +75,13 @@ Almost all file readers support accessing the data without reading it to memory
+The units of the navigation and signal axes can be converted automatically 
+during loading using the ``convert_units`` parameter. If `True`, the 
+``convert_to_units`` method of the ``axes_manager`` will be used for the conversion.
+If `None` (default), the default setting of the file reader will be used to 
```
I think that this part of the documentation needs updating to take into account the latest changes.

francisco-dlp reviewed (commented):
Will LGTM as soon as the little doc issue is fixed and Rnm merged in.

ericpre committed:
Merge remote-tracking branch 'upstream/RELEASE_next_minor' into scale_unit_conversion_take2

# Conflicts:
#	hyperspy/tests/drawing/plot_model/test_fit_EELS_convolved_False.png
#	hyperspy/tests/drawing/plot_model/test_fit_EELS_convolved_True.png
#	hyperspy/tests/drawing/plot_model/test_plot_gaussian_eelsmodel_False-False-False.png
#	hyperspy/tests/drawing/plot_model/test_plot_gaussian_eelsmodel_False-False-True.png
#	hyperspy/tests/drawing/plot_model/test_plot_gaussian_eelsmodel_False-True-False.png
#	hyperspy/tests/drawing/plot_model/test_plot_gaussian_eelsmodel_False-True-True.png
#	hyperspy/tests/drawing/plot_model/test_plot_gaussian_eelsmodel_True-False-False.png
#	hyperspy/tests/drawing/plot_model/test_plot_gaussian_eelsmodel_True-False-True.png
#	hyperspy/tests/drawing/plot_model/test_plot_gaussian_eelsmodel_True-True-False.png
#	hyperspy/tests/drawing/plot_model/test_plot_gaussian_eelsmodel_True-True-True.png
#	hyperspy/tests/drawing/plot_model1d/test_default_signal_plot.png
#	hyperspy/tests/drawing/plot_model1d/test_disable_plot_components.png
#	hyperspy/tests/drawing/plot_model1d/test_plot_components.png

ericpre committed:
Update user guide.

francisco-dlp commented:
unittests are failing

ericpre committed:
Update baseline image for failing tests following previous merge with upstream.

ericpre commented:
Sorry, usually I checked but I haven't this time... it should be fine now!
